lru/kv.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package lru
+
+import (
+	"container/list"
+	"sync"
+)
+
+// kv represents a key-value pair.
+type kv struct {
+	key   interface{}
+	value interface{}
+}
+
+// KVCache provides a concurrency safe least-recently-used key/value cache with
+// nearly O(1) lookups, inserts, and deletions.  The cache is limited to a
+// maximum number of items with eviction for the oldest entry when the
+// limit is exceeded.
+//
+// The NewKVCache function must be used to create a usable cache since the zero
+// value of this struct is not valid.
+type KVCache struct {
+	mtx   sync.Mutex
+	cache map[interface{}]*list.Element // nearly O(1) lookups
+	list  *list.List                    // O(1) insert, update, delete
+	limit uint
+}
+
+// Lookup returns the associated value of the passed key, if it is a member of
+// the cache. Looking up an existing item makes it the most recently used item.
+
+// This function is safe for concurrent access.
+func (m *KVCache) Lookup(key interface{}) (interface{}, bool) {
+	m.mtx.Lock()
+	node, exists := m.cache[key]
+	if exists {
+		m.list.MoveToFront(node)
+	}
+	m.mtx.Unlock()
+
+	if !exists {
+		return nil, exists
+	}
+
+	pair := node.Value.(kv)
+	return pair.value, exists
+}
+
+// Contains returns whether or not the passed key is a member of the cache.
+// The associated item of the passed key if it exists becomes the most
+// recently used item.
+//
+// This function is safe for concurrent access.
+func (m *KVCache) Contains(key interface{}) bool {
+	m.mtx.Lock()
+	node, exists := m.cache[key]
+	if exists {
+		m.list.MoveToFront(node)
+	}
+	m.mtx.Unlock()
+	return exists
+}
+
+// Add adds the passed k/v to the cache and handles eviction of the oldest pair
+// if adding the new pair would exceed the max limit.  Adding an existing pair
+// makes it the most recently used item.
+//
+// This function is safe for concurrent access.
+func (m *KVCache) Add(key interface{}, value interface{}) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	// When the limit is zero, nothing can be added to the cache, so just
+	// return.
+	if m.limit == 0 {
+		return
+	}
+
+	pair := kv{key: key, value: value}
```

Comment:
I'd make the pair a pointer to the `kv` instance and then reuse the existing pair instances to avoid extra allocations.  Since they're being boxed into an interface, they have to move to the heap anyway, so making it a pointer will just make it easier to update without causing extra allocations.

I would expect that to reduce the number of allocations to 2 in the benchmark as we had previously discussed a bit.

e.g.

```Go
if node, exists := m.cache[key]; exists {
	node.Value.(*kv).value = value
	m.list.MoveToFront(node)
	...
}

if uint(len(m.cache))+1 > m.limit {
	node := m.list.Back()
	lru := node.Value.(*kv)
...
	delete(m.cache, lru.key)
...
	lru.key = key
	lru.value = value
	m.list.MoveToFront(node)
...
}

// The limit hasn't been reached yet, so just add the new k/v.
node := m.list.PushFront(&kv{key: key, value: value})
m.cache[key] = node
```

Then of course you'd need to update the type asserting in `Lookup` too.

Reply:
Looking good. 
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ github.com/decred/dcrd/lru -bench ^(BenchmarkKV)$

goos: linux
goarch: amd64
pkg: github.com/decred/dcrd/lru
BenchmarkKV-8   	 3047961	       389 ns/op	      25 B/op	       2 allocs/op
PASS
ok  	github.com/decred/dcrd/lru	1.654s
Success: Benchmarks passed.
```